File: leetcode/Merge_Intervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */

class Solution {
public:
    vector<Interval> merge(vector<Interval> &intervals) {
        if(intervals.empty()) return vector<Interval>();
        
        sort(intervals.begin(), intervals.end(), Solution::interval_cmp);
        
        vector<Interval> result;
        int i = 1;
        int n = intervals.size();
        int s, e;
        
        s = intervals[0].start;
        e = intervals[0].end;
        while(i < n){
            if(intervals[i].start <= e){
                e = max(e, intervals[i].end);
            }else{
                result.push_back(Interval(s, e));
                s = intervals[i].start;
                e = intervals[i].end;
            }
            
            i++;
        }
        result.push_back(Interval(s, e));
        
        return result;
    }
private:
    static bool interval_cmp(const Interval &lhs, const Interval &rhs){
        return lhs.start < rhs.start;
    }
};

File: leetcode/java/MergeIntervals.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
class Solution {
    public List<Interval> merge(List<Interval> intervals) {
        if(intervals == null || intervals.isEmpty()) return Collections.emptyList();
        Collections.sort(intervals, (l, r) -> l.start - r.start);
        
        List<Interval> result = new ArrayList<>();
        int i = 0;
        int j = 1;
        while(j < intervals.size()) {
            Interval cur = intervals.get(i);
            Interval next = intervals.get(j);
            if(next.start <= cur.end) {
                cur.end = Math.max(cur.end, next.end);
            } else {
                result.add(cur);
                i = j;
            }
            j++;
        }
        result.add(intervals.get(i));
        return result;
    }
}
